lib/src/source/hosted.dart

```diff
@@ -537,6 +547,183 @@ class HostedSource extends CachedSource {
+        final pkg = element['package'];
+        if (pkg is! Map) {
+          throw FormatException('package must be a map');
+        }
+        final name = pkg['name'];
+        if (name is! String) {
+          throw FormatException('package name must be a String');
+        }
+        if (name == packageName) {
+          final ecosystem = pkg['ecosystem'];
+          if (ecosystem is! String) {
+            throw FormatException('ecosystem must be a String');
+          }
+          return ecosystem.toLowerCase() == 'pub';
+        }
+        return false;
+      }) as Map?;
+
+      if (affectedPkg == null) {
+        throw FormatException(
+          'Advisory $id does not contain $packageName among its affected packages.',
+        );
+      }
+
+      final versions = affectedPkg['versions'];
+      if (versions is! List) {
+        throw FormatException('package versions must be a list');
+      }
+
+      if (versions.any((element) => element is! String)) {
+        throw FormatException('package version elements must be a string');
+      }
+
+      advisoriesResult
+          .add(Advisory(id, [...versions.map((e) => e.toString())]));
+      affectedVersions[id] = [...versions.map((e) => e.toString())];
+    }
+
+    return advisoriesResult;
+  }
+
+  Future<List<Advisory>?> _getAdvisories(
+    PackageId id,
+    SystemCache cache,
+    Duration? maxAge,
+  ) async {
+    final advisoriesUpdated =
+        (await status(id.toRef(), id.version, cache, maxAge: maxAge))
+            .advisoriesUpdated;
+    if (advisoriesUpdated == null) return null;
+
+    Future<List<Advisory>?> readAdvisoriesFromCache() async {
+      final advisoriesCachePath = _advisoriesCachePath(id.toRef(), cache);
+      final stat = io.File(advisoriesCachePath).statSync();
+
+      if (stat.type == io.FileSystemEntityType.file) {
+        if (advisoriesUpdated.isAfter(stat.modified)) {
+          return null;
+        }
+
+        try {
+          final doc = jsonDecode(readTextFile(advisoriesCachePath));
+          if (doc is! Map) {
+            throw FormatException('Broken cached advisories response');
+          }
+
+          final cachedAdvisoriesUpdated = doc['advisoriesUpdated'];
+          if (cachedAdvisoriesUpdated is! String) {
+            throw FormatException('Broken cached advisories response');
+          }
+          final parsedCacheAdvisoriesUpdated =
+              DateTime.parse(cachedAdvisoriesUpdated);
+          if ((await status(id.toRef(), id.version, cache))
+              .advisoriesUpdated!
+              .isAfter(parsedCacheAdvisoriesUpdated)) {
+            // too old
+            tryDeleteEntry(advisoriesCachePath);
+          } else {
+            return _parseAdvisories(doc, id.toRef().name);
+          }
+        } on io.IOException {
+          // Could not read the file. Delete if it exists.
+          tryDeleteEntry(advisoriesCachePath);
+        } on FormatException {
+          tryDeleteEntry(advisoriesCachePath);
```

Comment:
Perhaps do a `log.fine('Failed to read cached version listing: $e');`

Reply:
done